Assets/Scripts/Gameplay/ConnectionManagement/ConnectionState/OfflineState.cs

```diff
@@ -0,0 +1,197 @@
+        {
+            if (await JoinRelayServerAsync(onFailure))
+            {
+                ConnectClient(GetPlayerId(), playerName);
+            }
+        }
+
+        protected async Task<bool> JoinRelayServerAsync(Action<string> onFailure)
+        {
+            Debug.Log($"Setting Unity Relay client with join code {m_LocalLobby.RelayJoinCode}");
+
+            try
+            {
+                var (ipv4Address, port, allocationIdBytes, connectionData, hostConnectionData, key) =
+                    await UnityRelayUtilities.JoinRelayServerFromJoinCode(m_LocalLobby.RelayJoinCode);
+
+                await m_LobbyServiceFacade.UpdatePlayerRelayInfoAsync(allocationIdBytes.ToString(), m_LocalLobby.RelayJoinCode);
+                var utp = (UnityTransport)m_ConnectionManager.NetworkManager.NetworkConfig.NetworkTransport;
+                utp.SetClientRelayData(ipv4Address, port, allocationIdBytes, key, connectionData, hostConnectionData, isSecure: true);
+            }
+            catch (Exception e)
+            {
+                onFailure?.Invoke(e.Message);
+                return false;
+            }
+
+            return true;
+        }
+
+        public override bool StartHostIP(string playerName, string ipaddress, int port)
+        {
+            var utp = (UnityTransport)NetworkManager.Singleton.NetworkConfig.NetworkTransport;
+            utp.SetConnectionData(ipaddress, (ushort)port);
+
+            return StartHost(GetPlayerId(), playerName);
+        }
+
+        public override async Task StartHostLobbyAsync(string playerName)
+        {
+            Debug.Log("Setting up Unity Relay host");
+
+            try
+            {
+                var (ipv4Address, port, allocationIdBytes, connectionData, key, joinCode) =
+                    await UnityRelayUtilities.AllocateRelayServerAndGetJoinCode(k_MaxUnityRelayConnections);
+
+                m_LocalLobby.RelayJoinCode = joinCode;
+                //next line enabled lobby and relay services integration
+                await m_LobbyServiceFacade.UpdateLobbyDataAsync(m_LocalLobby.GetDataForUnityServices());
+                await m_LobbyServiceFacade.UpdatePlayerRelayInfoAsync(allocationIdBytes.ToString(), joinCode);
+
+                // we now need to set the RelayCode somewhere :P
+                var utp = (UnityTransport)m_ConnectionManager.NetworkManager.NetworkConfig.NetworkTransport;
+                utp.SetHostRelayData(ipv4Address, port, allocationIdBytes, key, connectionData, isSecure: true);
+            }
+            catch (Exception e)
+            {
+                Debug.LogErrorFormat($"{e.Message}");
+                throw;
+            }
+
+            StartHost(GetPlayerId(), playerName);
+        }
+
+        public override void ApprovalCheck(byte[] connectionData, ulong clientId, NetworkManager.ConnectionApprovedDelegate connectionApprovedCallback)
```

Comment:
oooh that's interesting that now that we have this pattern, our approval is separated in two calls... mmm it seems a bit weird though that approval happens while you're offline, couldn't this be in a "starting host" state?

Reply:
Done, it's much cleaner now I think